### PR TITLE
:bug:: Fix controller-pod SA `permission denied` error

### LINKED ISSUE #3016: Got errors on kubebuilder v3 quick start
### What broke? What's expected?

Using the latest kubebuilder cli, quick start a new project based on the [official document](https://book.kubebuilder.io/quick-start.html), got these errors:
1. Enable the leader-election, then `make run`, got error:
```
go fmt ./...
go vet ./...
go run ./main.go
1.665666448465308e+09   ERROR   setup   unable to start manager {"error": "unable to find leader election namespace: not running in-cluster, please specify LeaderElectionNamespace"}
main.main
...
exit status 1
make: *** [run] Error 1
```

2. After `make deploy IMG=xxx/xxx:tag`, the controller-pod got error:
```                                    
1.6656687258729894e+09  ERROR   controller-runtime.client.config        unable to get kubeconfig        {"error": "open /var/run/secrets/kubernetes.io/serviceaccount/token: permission denied"}
sigs.k8s.io/controller-runtime/pkg/client/config.GetConfigOrDie
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/client/config/config.go:153
main.main
        /workspace/main.go:68
runtime.main
        /usr/local/go/src/runtime/proc.go:250
```

3. As the doc said `Important: Run "make" to regenerate code after modifying this file`, but `make` did not generate latest CRD fields.

### Reproducing this issue

Just quick start a new project based on the [official document](https://book.kubebuilder.io/quick-start.html), then will reproduce this issue.

### KubeBuilder (CLI) Version

Version: main.version{KubeBuilderVersion:"3.7.0", KubernetesVendor:"1.24.1", GitCommit:"3bfc84ec8767fa760d1771ce7a0cb05a9a8f6286", BuildDate:"2022-09-20T17:21:57Z", GoOs:"darwin", GoArch:"amd64"}

### PROJECT version

_No response_

### Plugin versions

_No response_

### Other versions

_No response_

### Extra Labels

_No response_

### LINKED ISSUE #3028: Default v3 scaffold manager pod got permission denied error
### What broke? What's expected?

Based on the quick start doc, did the `make manifests, make install`, then
`make docker-build docker-push IMG=xxx/xxx:tag, make deploy IMG=xxx/xxx:tag`, then
in the cluster the controller-pod got error:
```
1.6656687258729894e+09  ERROR   controller-runtime.client.config        unable to get kubeconfig        {"error": "open /var/run/secrets/kubernetes.io/serviceaccount/token: permission denied"}
sigs.k8s.io/controller-runtime/pkg/client/config.GetConfigOrDie
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/client/config/config.go:153
main.main
        /workspace/main.go:68
runtime.main
        /usr/local/go/src/runtime/proc.go:250
```

### Reproducing this issue

Based on the quick start doc, did the `make manifests, make install`, then
`make docker-build docker-push IMG=xxx/xxx:tag, make deploy IMG=xxx/xxx:tag`, then
in the cluster will reproduce this issue.

My cluster K8s version is:
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:31:21Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:25:06Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
```
Kubebuilder version is:
```
Version: main.version{KubeBuilderVersion:"3.7.0", KubernetesVendor:"1.24.1", GitCommit:"3bfc84ec8767fa760d1771ce7a0cb05a9a8f6286", BuildDate:"2022-09-20T17:21:57Z", GoOs:"darwin", GoArch:"amd64"}
```
With these context, will replicate the error.

### KubeBuilder (CLI) Version

Version: main.version{KubeBuilderVersion:"3.7.0", KubernetesVendor:"1.24.1", GitCommit:"3bfc84ec8767fa760d1771ce7a0cb05a9a8f6286", BuildDate:"2022-09-20T17:21:57Z", GoOs:"darwin", GoArch:"amd64"}

### PROJECT version

_No response_

### Plugin versions

_No response_

### Other versions

_No response_

### Extra Labels

_No response_

### PR DESCRIPTION
**Pull request by astraw99**
Fixes #3028

Add `fsGroup` based on [dockerfile.go](https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/golang/v3/scaffolds/internal/templates/dockerfile.go#L71) to get proper access to open `serviceaccount` token file.

### CODE REVIEW

File: pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go [hunk 1]
```diff
@@ -104,6 +104,7 @@ spec:
       #             values:
       #               - linux
       securityContext:
+        fsGroup: 65532
```
Comment (camilamacedo86) [inline pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go:107]: /hold

Why are you adding the fsGroup here? 
It is not required.

File: pkg/plugins/golang/v3/scaffolds/internal/templates/main.go [hunk 2]
```diff
@@ -239,12 +239,13 @@ func main() {
...
-		Port:                   9443,
-		HealthProbeBindAddress: probeAddr,
-		LeaderElection:         enableLeaderElection,
-		LeaderElectionID:       "{{ hashFNV .Repo }}.{{ .Domain }}",
+		Scheme:                  scheme,
+		MetricsBindAddress:      metricsAddr,
+		Port:                    9443,
+		HealthProbeBindAddress:  probeAddr,
+		LeaderElection:          enableLeaderElection,
+		LeaderElectionNamespace: "{{ .ProjectName }}-system",
```
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/main.go:247]: We have more than one change/motivation here. 
Could you please create a PR for each case scenario? 
Also, can you please add a description of the issue faced to let us know why would you like to propose the change?

Note that the title of the pr is used to generate the changelogs so would be nice to have something that better clarifies the reason for the changes.

File: pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go [hunk 3]
```diff
@@ -121,7 +121,7 @@ test: manifests generate fmt vet envtest ## Run tests.
 ##@ Build
 
 .PHONY: build
-build: generate fmt vet ## Build manager binary.
+build: manifests generate fmt vet ## Build manager binary.
```
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go:124]: It seems a nice change.
We would need to do that for go/v3 and go/v4-alpha plugins

Could you please create a PR for this one?

Comment (astraw99) [reply pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go:124]: Got it, raised a new PR: https://github.com/kubernetes-sigs/kubebuilder/pull/3020
PTAL thanks.

File: pkg/plugins/golang/v3/scaffolds/internal/templates/main.go (hunk 2 above)
Comment (astraw99) [reply pkg/plugins/golang/v3/scaffolds/internal/templates/main.go:247]: Got it, raised a new PR: https://github.com/kubernetes-sigs/kubebuilder/pull/3021
PTAL thanks.

File: pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go (hunk 1 above)
Comment (astraw99) [reply pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go:107]: Described in issue #3016, the controller-pod start got error:
```
1.6656687258729894e+09  ERROR   controller-runtime.client.config        unable to get kubeconfig        {"error": "open /var/run/secrets/kubernetes.io/serviceaccount/token: permission denied"}
```
So add fsGroup (based on [dockerfile.go](https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/golang/v3/scaffolds/internal/templates/dockerfile.go#L71)) to resolve it.

Comment (astraw99) [reply pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go:107]: Maybe I did not make it clear in issue #3016, raised a new issue #3028 to clarify it.
PTAL thanks.

Comment (camilamacedo86) [reply pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go:107]: Hi @astraw99, 

> Only faced the sa permission error, opened a new issue https://github.com/kubernetes-sigs/kubebuilder/issues/3028 to clarify it.

The default scaffold does have not SA. So, how can that can be faced by default? 
What are the steps required to be performed from scratch so we are able to reproduce this issue?

File: pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go (hunk 3 above)
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go:124]: could you revert this change? It was done in another PR

Comment (astraw99) [reply pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go]: Done.

File: pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go (hunk 1 above)
Comment (astraw99) [reply pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go]: As said in https://github.com/kubernetes-sigs/kubebuilder/issues/3028#issuecomment-1287849547, here add a `fsGroup` comment doc, to make it helpful to resolve the issue.
PTAL thanks.

File: pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go [hunk 4]
```diff
@@ -105,6 +105,8 @@ spec:
       #               - linux
       securityContext:
         runAsNonRoot: true
+        # if your K8s version <= 1.21, may need to uncomment the follow fsGroup to get SA token file access.
```
Comment (camilamacedo86) [inline pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go:108]: Hi @astraw99,

I do not think that we should add/change the scaffold.
(IHMO)

a) It could be fit in an issue as a workaround (so that others can find the problem and solution)
b) We could add it in the docs under a new FAQ section  
c) It would be great better understand it and make clear that ONLY happens in development env when we are running/testing the project OUTSIDE of the cluster with make run.

Comment (camilamacedo86) [reply pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go:108]: /hold

Comment (astraw99) [reply pkg/plugins/common/kustomize/v2-alpha/scaffolds/internal/templates/config/manager/config.go:108]: OK, maybe it is more proper to add this case into the FAQ section (I am doing).

**Review by camilamacedo86 — COMMENTED** [review]
Hi @astraw99,

The issue does not occur as you checked in the latest version of k8s.
So, I do not think that we should change the scaffolds with a workaround to solve issues for old k8s versions.
Also, we could address this scenario with the FAQ doc: https://github.com/kubernetes-sigs/kubebuilder/pull/3044

So, wdyt? Could we close this one? Is that make sense?

**Comment by astraw99** [comment]
OK, we can close it now.
